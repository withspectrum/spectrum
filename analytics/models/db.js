// @flow
/**
 * Database setup is done here
 */
const fs = require('fs');
const path = require('path');
const IS_PROD = !process.env.FORCE_DEV && process.env.NODE_ENV === 'production';

const DEFAULT_CONFIG = {
  // Connect to the test database when, well, testing
  db: !process.env.TEST_DB ? 'spectrum' : 'testing',
  max: 500, // Maximum number of connections, default is 1000
  buffer: 5, // Minimum number of connections open at any given moment, default is 50
  timeoutGb: 60 * 1000, // How long should an unused connection stick around, default is an hour, this is a minute
};

const PRODUCTION_CONFIG = {
  password: process.env.COMPOSE_US_WEST_2_RETHINKDB_PASSWORD,
  user: 'admin',
  host: process.env.COMPOSE_US_WEST_2_RETHINKDB_URL,
  port: process.env.COMPOSE_US_WEST_2_RETHINKDB_PORT,
  ssl: {
    ca: fs.readFileSync(path.join(process.cwd(), 'cacert')),
  },
};

const config = IS_PROD
  ? {
      ...DEFAULT_CONFIG,
      ...PRODUCTION_CONFIG,
    }
  : {
      ...DEFAULT_CONFIG,
    };

var r = require('rethinkdbdash')(config);

module.exports = { db: r };
