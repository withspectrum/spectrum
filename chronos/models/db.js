/**
 * Database setup is done here
 */
const fs = require('fs');
const path = require('path');
const IS_PROD = !process.env.FORCE_DEV && process.env.NODE_ENV === 'production';

const DEFAULT_CONFIG = {
  db: 'spectrum',
};

const PRODUCTION_CONFIG = {
  password: process.env.AWS_RETHINKDB_PASSWORD,
  host: process.env.AWS_RETHINKDB_URL,
  port: process.env.AWS_RETHINKDB_PORT,
};

const config = !IS_PROD
  ? {
      ...DEFAULT_CONFIG,
      ...PRODUCTION_CONFIG,
    }
  : {
      ...DEFAULT_CONFIG,
    };

var r = require('rethinkdbdash')(config);

module.exports = { db: r };
