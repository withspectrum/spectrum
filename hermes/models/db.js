// @flow
const IS_PROD = !process.env.FORCE_DEV && process.env.NODE_ENV === 'production';

const DEFAULT_CONFIG = {
  db: 'spectrum',
  max: 100, // Maximum number of connections, default is 1000
  buffer: 10, // Minimum number of connections open at any given moment, default is 50
  timeoutGb: 60 * 1000, // How long should an unused connection stick around, default is an hour, this is a minute
};

const PRODUCTION_CONFIG = {
  password: process.env.AWS_RETHINKDB_PASSWORD,
  host: process.env.AWS_RETHINKDB_URL,
  port: process.env.AWS_RETHINKDB_PORT,
};

const config = IS_PROD
  ? {
      ...DEFAULT_CONFIG,
      ...PRODUCTION_CONFIG,
    }
  : {
      ...DEFAULT_CONFIG,
    };

const r = require('rethinkdbdash')(config);

module.exports = { db: r };
