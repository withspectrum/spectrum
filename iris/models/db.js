// @flow
/**
 * Database setup is done here
 */
const IS_PROD = !process.env.FORCE_DEV && process.env.NODE_ENV === 'production';

const DEFAULT_CONFIG = {
  // Connect to the test database when, well, testing
  db: !process.env.TEST_DB ? 'spectrum' : 'testing',
  max: 500, // Maximum number of connections, default is 1000
  buffer: 5, // Minimum number of connections open at any given moment, default is 50
  timeoutGb: 60 * 1000, // How long should an unused connection stick around, default is an hour, this is a minute
  pingInterval: 300, // Ping the connection every 5 minutes (300 seconds) to keep it alive and prevent rethinkdbdash#192
};

const PRODUCTION_CONFIG = {
  password: process.env.AWS_RETHINKDB_PASSWORD,
  host: process.env.AWS_RETHINKDB_URL,
  port: process.env.AWS_RETHINKDB_PORT,
};

const config = IS_PROD
  ? {
      ...DEFAULT_CONFIG,
      ...PRODUCTION_CONFIG,
    }
  : {
      ...DEFAULT_CONFIG,
    };

var r = require('rethinkdbdash')(config);

if (process.env.NODE_ENV === 'development') {
  const fs = require('fs');
  const inspect = require('rethinkdb-inspector');
  const queries = [];
  inspect(r, {
    onQueryComplete: (query, { size, time }) => {
      if (query.indexOf('.changes') > -1) return;
      queries.push({ query, time, size });
      fs.writeFileSync(
        'queries-by-time.js',
        JSON.stringify(queries.sort((a, b) => b.time - a.time), null, 2)
      );
      fs.writeFileSync(
        'queries-by-response-size.js',
        JSON.stringify(queries.sort((a, b) => b.size - a.size), null, 2)
      );
    },
  });
}

module.exports = { db: r };
