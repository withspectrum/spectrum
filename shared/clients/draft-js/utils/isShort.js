// @flow
// Determine whether a DraftJS message is short
import { toPlainText, toState } from '../../../draft-utils';

import type { RawDraftContentState } from 'draft-js';
import type { MessageInfoType } from '../../../graphql/fragments/message/messageInfo';

export const isShort = (message: MessageInfoType): boolean => {
  if (message.messageType === 'media') return false;
  const jsonBody = JSON.parse(message.content.body);
  return (
    jsonBody.blocks.length <= 1 && toPlainText(toState(jsonBody)).length <= 170
  );
};
