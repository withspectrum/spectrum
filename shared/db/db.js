// @flow
/**
 * Database setup is done here
 */
import fs from 'fs';
import path from 'path';
import inspect from 'rethinkdb-inspector';
import { statsd } from '../statsd';

const IS_PROD = !process.env.FORCE_DEV && process.env.NODE_ENV === 'production';

const CONNECTIONS = 20;
const DEFAULT_CONFIG = {
  host: process.env.COMPOSE_RETHINKDB_URL,
  // Connect to the test database when, well, testing
  db: !process.env.TEST_DB ? 'spectrum' : 'testing',
  max: CONNECTIONS, // Maximum number of connections, default is 1000
  buffer: CONNECTIONS, // Minimum number of connections open at any given moment, default is 50
  timeoutGb: 60 * 60 * 1000, // How long should an unused connection stick around, default is an hour, this is a minute
  timeout: 30, // The number of seconds for a connection to be opened, default 20
};

let ca;

try {
  ca = fs.readFileSync(path.join(process.cwd(), 'cacert'));
} catch (err) {}

if (!ca && IS_PROD)
  throw new Error(
    'Please provide the SSL certificate to connect to the production database in a file called `cacert` in the root directory.'
  );

const PRODUCTION_CONFIG = {
  password: process.env.COMPOSE_RETHINKDB_PASSWORD,
  host: process.env.COMPOSE_RETHINKDB_URL,
  port: process.env.COMPOSE_RETHINKDB_PORT,
  ...(ca
    ? {
        ssl: {
          ca,
        },
      }
    : {}),
};

const config = IS_PROD
  ? {
      ...DEFAULT_CONFIG,
      ...PRODUCTION_CONFIG,
    }
  : {
      ...DEFAULT_CONFIG,
    };

let r = require('rethinkhaberdashery')(config);
const poolMaster = r.getPoolMaster();

poolMaster.on('queueing', size => {
  statsd.gauge('db.query_queue.size', size);
});

setInterval(() => {
  statsd.gauge('db.connections.count', poolMaster.getLength());
}, 5000);

// Exit the process on unhealthy db in test env
if (process.env.TEST_DB) {
  poolMaster.on('healthy', healthy => {
    if (!healthy) {
      process.exit(1);
    }
  });
}

const queries = [];
inspect(r, {
  onQueryComplete: (query, { size, time }) => {
    if (query.indexOf('.changes') > -1) return;
    statsd.increment('db.queries.count');
    statsd.histogram('db.queries.response_time', time);
    statsd.histogram('db.queries.response_size', size);
    // In development write out a file of the most expensive queries
    if (process.env.NODE_ENV === 'development' && process.env.TRACK_DB_PERF) {
      queries.push({ query, time, size });
      fs.writeFileSync(
        'queries-by-time.js',
        JSON.stringify(queries.sort((a, b) => b.time - a.time), null, 2)
      );
      fs.writeFileSync(
        'queries-by-response-size.js',
        JSON.stringify(queries.sort((a, b) => b.size - a.size), null, 2)
      );
    }
  },
});

module.exports = { db: r };
