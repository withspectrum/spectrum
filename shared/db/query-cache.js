// @flow
import TagCache from 'redis-tag-cache';

const IS_PROD = process.env.NODE_ENV === 'production' && !process.env.FORCE_DEV;

const DEFAULT_REDIS_OPTIONS = {
  keyPrefix: 'query-cache',
  host: process.env.REDIS_CACHE_URL,
};

const PRODUCTION_REDIS_OPTIONS = {
  port: process.env.REDIS_CACHE_PORT,
  host: process.env.REDIS_CACHE_URL,
  password: process.env.REDIS_CACHE_PASSWORD,
};

const queryCache = new TagCache({
  defaultTimeout: 86400,
  redis: {
    ...DEFAULT_REDIS_OPTIONS,
    ...(IS_PROD ? PRODUCTION_REDIS_OPTIONS : {}),
  },
});

export default queryCache;
