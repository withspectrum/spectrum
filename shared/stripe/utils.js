// @flow
const debug = require('debug')('pluto:queues:stripe-util');
import { stripe } from 'shared/stripe';
import type { RawCustomer } from 'shared/stripe/types/customer';
import type { RawSubscription } from 'shared/stripe/types/subscription';
import type { DBCommunity } from 'shared/types';
import type { RawSubscriptionItem } from 'shared/stripe/types/subscriptionItem';
import type { RawSource } from 'shared/stripe/types/source';
import type { RawInvoice } from 'shared/stripe/types/invoice';
import { getCommunityById, setStripeCustomerId } from 'api/models/community';
import {
  FREE_MODERATOR_SEAT,
  MODERATOR_SEAT,
  FREE_PRIVATE_CHANNEL,
  PRIVATE_CHANNEL,
  COMMUNITY_FEATURES,
  COMMUNITY_ANALYTICS,
} from 'pluto/queues/constants';

const getCustomer = async (customerId: string): Promise<RawCustomer> => {
  return await stripe.customers.retrieve(customerId);
};

const hasChargeableSource = (customer: RawCustomer): boolean => {
  if (!customer) return false;
  if (!customer.sources || customer.sources.data.length === 0) return false;
  return customer.sources.data.some(
    source => source && source.status === 'chargeable'
  );
};

const getChargeableSource = (customer: RawCustomer): ?RawSource => {
  if (!customer) return null;
  if (!customer.sources || customer.sources.data.length === 0) return null;
  return customer.sources.data.find(
    source => source && source.status === 'chargeable'
  );
};

const getSources = (
  customer: RawCustomer
): Array<?{
  ...$Exact<RawSource>,
  isDefault: boolean,
}> => {
  if (!customer) return [];
  if (!customer.sources || customer.sources.data.length === 0) return [];

  return customer.sources.data.map(source => {
    if (!source) return null;
    return Object.assign({}, source, {
      isDefault: source.id === customer.default_source,
    });
  });
};

type AttachNewSourceInput = {
  customerId: string,
  sourceId: string,
};
// prettier-ignore
const attachNewSource = async (input: AttachNewSourceInput): Promise<RawSource> => {
  const { customerId, sourceId } = input;
  return await stripe.customers.createSource(customerId, {
    source: sourceId,
  });
};

type DetachSourceInput = {
  customerId: string,
  sourceId: string,
};
const detachSource = async (input: DetachSourceInput): Promise<RawCustomer> => {
  const { customerId, sourceId } = input;
  return await stripe.customers.deleteSource(customerId, sourceId);
};

type ChangeDefaultSourceInput = {
  customerId: string,
  sourceId: string,
};
// prettier-ignore
const changeDefaultSource = async (input: ChangeDefaultSourceInput): Promise<RawCustomer> => {
  const { customerId, sourceId } = input;
  return await stripe.customers.update(customerId, {
    default_source: sourceId,
  });
};

const hasActiveSubscription = (customer: RawCustomer): boolean => {
  if (!customer) return false;
  if (!customer.subscriptions || customer.subscriptions.data.length === 0)
    return false;
  return customer.subscriptions.data.some(
    sub => sub && sub.status === 'active'
  );
};

const getActiveSubscription = (customer: RawCustomer): ?RawSubscription => {
  if (!customer) return null;
  if (!customer.subscriptions || customer.subscriptions.data.length === 0)
    return null;
  return customer.subscriptions.data.find(
    sub => sub && sub.status === 'active'
  );
};

const getSubscriptions = (customer: RawCustomer): Array<?RawSubscription> => {
  if (!customer) return [];
  if (!customer.subscriptions || customer.subscriptions.data.length === 0)
    return [];
  return customer.subscriptions.data;
};

const hasSubscriptionItemOfType = (
  customer: RawCustomer,
  type: string
): boolean => {
  if (!customer) return false;
  const activeSubscription = getActiveSubscription(customer);
  if (!activeSubscription) return false;
  return activeSubscription.items.data.some(
    item => item && item.plan.id === type
  );
};

type SubscriptionItemTypes =
  | typeof FREE_MODERATOR_SEAT
  | typeof MODERATOR_SEAT
  | typeof FREE_PRIVATE_CHANNEL
  | typeof PRIVATE_CHANNEL
  | typeof COMMUNITY_ANALYTICS
  | typeof COMMUNITY_FEATURES;

// prettier-ignore
const getSubscriptionItemOfType = (customer: RawCustomer, type: SubscriptionItemTypes): ?RawSubscriptionItem => {
  if (!customer) return null;
  const activeSubscription = getActiveSubscription(customer);
  if (!activeSubscription) return null;
  return activeSubscription.items.data.find(
    item => item && item.plan.id === type
  );
};

type CreateCustomerInput = {
  administratorEmail: ?string,
  communityId: string,
  communityName: string,
};
type CreateCustomerOutput = {
  customer: RawCustomer,
  community: DBCommunity,
};
// prettier-ignore
const createCustomer = async (customerInput: CreateCustomerInput): Promise<CreateCustomerOutput> => {
  const customer = await stripe.customers.create({
    email: customerInput.administratorEmail,
    metadata: {
      communityId: customerInput.communityId,
      communityName: customerInput.communityName,
    },
  });

  const community = await setStripeCustomerId(
    customerInput.communityId,
    customer.id
  );

  return {
    customer,
    community,
  };
};

const deleteCustomer = async (customerId: string) => {
  return await stripe.customers.del(customerId);
};

type UpdateCustomerInput = {
  customerId: string,
  administratorEmail: ?string,
  communityId: string,
  communityName: string,
};
const updateCustomer = async (customerInput: UpdateCustomerInput) => {
  return await stripe.customers.update(customerInput.customerId, {
    email: customerInput.administratorEmail,
    metadata: {
      communityId: customerInput.communityId,
      communityName: customerInput.communityName,
    },
  });
};

type FirstSubInput = {
  customerId: string,
  subscriptionItemType: SubscriptionItemTypes,
};
const createFirstSubscription = async (input: FirstSubInput) => {
  const { customerId, subscriptionItemType } = input;
  return await stripe.subscriptions.create({
    customer: customerId,
    items: [
      // NOTE: We have to include this dummy item in order to prevent
      // the top-level subscription from thinking it's about any
      // specific feature
      {
        plan: COMMUNITY_FEATURES,
        quantity: 1,
      },
      {
        plan: subscriptionItemType,
        quantity: 1,
      },
    ],
  });
};

// given subscriptions returned from a StripeCustomers record in our db,
// we reformat the data to be more front-end friendly
const cleanSubscriptions = (customer: RawCustomer): Array<?Object> => {
  const subscriptions = getSubscriptions(customer);
  if (!subscriptions || subscriptions.length === 0) return [];

  const formatItems = (items: Array<any>) => {
    if (!items || items.length === 0) return [];
    return items
      .filter(item => item && item.plan.id !== COMMUNITY_FEATURES)
      .map(
        item =>
          item && {
            created: item.created,
            planId: item.plan.id,
            planName: item.plan.name,
            amount: item.plan.amount,
            quantity: item.quantity,
            id: item.id,
          }
      );
  };

  return subscriptions.map(
    sub =>
      sub && {
        id: sub.id,
        created: sub.created,
        current_period_end: sub.current_period_end,
        discount: sub.discount
          ? {
              amount_off: sub.discount.coupon.amount_off,
              percent_off: sub.discount.coupon.percent_off,
              id: sub.discount.coupon.id,
            }
          : customer.discount
            ? {
                amount_off: customer.discount.coupon.amount_off,
                percent_off: customer.discount.coupon.percent_off,
                id: customer.discount.coupon.id,
              }
            : null,
        billing_cycle_anchor: sub.billing_cycle_anchor,
        canceled_at: sub.canceled_at,
        items: formatItems(sub.items.data),
        status: sub.status,
      }
  );
};

const cleanInvoices = (invoices: Array<?RawInvoice>) => {
  if (!invoices || invoices.length === 0) return [];

  const cleanInvoiceItems = items => {
    return items.map(
      item =>
        item && {
          id: item.id,
          amount: item.plan.amount,
          quantity: item.quantity,
          planId: item.plan.id,
          planName: item.plan.name,
        }
    );
  };

  return invoices
    .map(
      invoice =>
        invoice && {
          id: invoice.id,
          date: invoice.date,
          items: cleanInvoiceItems(invoice.lines.data),
          total: invoice.total,
        }
    )
    .filter(invoice => invoice && invoice.total > 0);
};

type AddSIInput = {
  subscriptionId: string,
  subscriptionItemType: SubscriptionItemTypes,
};
const addSubscriptionItem = async (input: AddSIInput) => {
  const { subscriptionId, subscriptionItemType } = input;

  return await stripe.subscriptionItems.create({
    subscription: subscriptionId,
    plan: subscriptionItemType,
    quantity: 1,
    prorate: true,
  });
};

type UpdateSIInput = {
  subscriptionItemId: string,
  quantity: number,
};
const updateSubscriptionItem = async (input: UpdateSIInput) => {
  const { subscriptionItemId, quantity } = input;

  return await stripe.subscriptionItems.update(subscriptionItemId, {
    prorate: true,
    quantity,
  });
};

const deleteSubscriptionItem = async (subscriptionItemId: string) => {
  return await stripe.subscriptionItems.del(subscriptionItemId);
};

type PreflightCheck = {
  community: ?DBCommunity,
  customer: ?RawCustomer,
  activeSubscription: ?RawSubscription,
  hasChargeableSource: boolean,
};
const jobPreflight = async (communityId: string): Promise<PreflightCheck> => {
  let defaultResult = {
    community: null,
    customer: null,
    activeSubscription: null,
    hasChargeableSource: false,
  };

  if (!communityId) {
    debug('No communityId in jobPreflight');
    return await defaultResult;
  }

  let dbCommunity = await getCommunityById(communityId);
  if (!dbCommunity) {
    debug('No community in db found during jobPreflight');
    return await defaultResult;
  }

  const { stripeCustomerId, administratorEmail, name, id } = dbCommunity;

  let customer;
  if (stripeCustomerId) {
    debug('Fetching customer from stripe in jobPreflight');
    try {
      customer = await getCustomer(stripeCustomerId);
    } catch (err) {
      return defaultResult;
    }
  } else {
    debug('Creating customer on stripe in jobPreflight');
    const createdCustomer = await createCustomer({
      administratorEmail,
      communityId: id,
      communityName: name,
    });
    customer = createdCustomer.customer;
    // if a customer was created in the preflight, always make sure to return
    // the updated community with the stripeCustomerId field set
    dbCommunity = await getCommunityById(communityId);
  }

  if (!customer) return defaultResult;

  debug('Preflight check completing...');

  return {
    customer,
    community: dbCommunity,
    activeSubscription: getActiveSubscription(customer),
    hasChargeableSource: hasChargeableSource(customer),
  };
};

export const StripeUtil = {
  getCustomer,
  hasChargeableSource,
  getChargeableSource,
  getSources,
  attachNewSource,
  detachSource,
  changeDefaultSource,
  hasActiveSubscription,
  getActiveSubscription,
  getSubscriptions,
  hasSubscriptionItemOfType,
  getSubscriptionItemOfType,
  cleanSubscriptions,
  cleanInvoices,
  createCustomer,
  deleteCustomer,
  updateCustomer,
  createFirstSubscription,
  addSubscriptionItem,
  updateSubscriptionItem,
  deleteSubscriptionItem,
  jobPreflight,
};
