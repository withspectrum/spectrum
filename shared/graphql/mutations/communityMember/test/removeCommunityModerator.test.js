//@flow
import { request } from '../../../../../iris/test/utils';
import db from 'shared/testing/db';
import data from 'shared/testing/data';

const owner = data.users.find(({ username }) => username === 'mxstbr');
const moderator = data.users.find(({ username }) => username === 'brian');
const member = data.users.find(({ username }) => username === 'bryn');
const nonMember = data.users.find(({ username }) => username === 'bad-boy');
const blockedMember = data.users.find(
  ({ username }) => username === 'blocked-boy'
);

const input = { communityId: data.communities[0].id, userId: moderator.id };
const variables = { input };

const dropUsersCommunities = () =>
  db
    .table('usersCommunities')
    .delete()
    .run();

const populateUsersCommunities = () =>
  db
    .table('usersCommunities')
    .insert(...data.usersCommunities)
    .run();

afterAll(
  async () =>
    await dropUsersCommunities().then(
      async () => await populateUsersCommunities()
    )
);

it('should fail if current user is not authenticated', async () => {
  const query = /* GraphQL */ `
    mutation removeCommunityModerator($input: RemoveCommunityModeratorInput!) {
      removeCommunityModerator (input: $input) {
        id
      }
    }
  `;

  expect.assertions(2);
  const result = await request(query, { variables });
  expect(result.errors).toHaveLength(1);
  expect(result.data.removeCommunityModerator).toEqual(null);
});

it('should fail if current user is not a member of the community', async () => {
  const query = /* GraphQL */ `
    mutation removeCommunityModerator($input: RemoveCommunityModeratorInput!) {
      removeCommunityModerator (input: $input) {
        id
      }
    }
  `;

  const context = {
    user: nonMember,
  };

  expect.assertions(2);
  const result = await request(query, { context, variables });
  expect(result.errors).toHaveLength(1);
  expect(result.data.removeCommunityModerator).toEqual(null);
});

it('should fail if current user is not an owner of the community', async () => {
  const query = /* GraphQL */ `
    mutation removeCommunityModerator($input: RemoveCommunityModeratorInput!) {
      removeCommunityModerator (input: $input) {
        id
      }
    }
  `;

  const context = {
    user: member,
  };

  expect.assertions(2);
  const result = await request(query, { context, variables });
  expect(result.errors).toHaveLength(1);
  expect(result.data.removeCommunityModerator).toEqual(null);
});

it('should fail if evaluated user is not a member of the community', async () => {
  const query = /* GraphQL */ `
    mutation removeCommunityModerator($input: RemoveCommunityModeratorInput!) {
      removeCommunityModerator (input: $input) {
        id
      }
    }
  `;

  const context = {
    user: owner,
  };

  const thisInput = {
    communityId: data.communities[0].id,
    userId: nonMember.id,
  };
  const thisVariables = { input: thisInput };

  expect.assertions(2);
  const result = await request(query, { context, variables: thisVariables });
  expect(result.errors).toHaveLength(1);
  expect(result.data.removeCommunityModerator).toEqual(null);
});

it("should fail if the community doesn't exist", async () => {
  const query = /* GraphQL */ `
    mutation removeCommunityModerator($input: RemoveCommunityModeratorInput!) {
      removeCommunityModerator (input: $input) {
        id
      }
    }
  `;

  const thisInput = { communityId: 'blah', userId: moderator.id };
  const thisVariables = { input: thisInput };

  const context = {
    user: owner,
  };

  expect.assertions(2);
  const result = await request(query, { context, variables: thisVariables });
  expect(result.errors).toHaveLength(1);
  expect(result.data.removeCommunityModerator).toEqual(null);
});

it('should fail if evaluated user not a moderator in the community', async () => {
  const query = /* GraphQL */ `
    mutation removeCommunityModerator($input: RemoveCommunityModeratorInput!) {
      removeCommunityModerator (input: $input) {
        id
      }
    }
  `;

  const thisInput = {
    communityId: data.communities[0].id,
    userId: member.id,
  };
  const thisVariables = { input: thisInput };

  const context = {
    user: owner,
  };

  expect.assertions(2);
  const result = await request(query, { context, variables: thisVariables });
  expect(result.errors).toHaveLength(1);
  expect(result.data.removeCommunityModerator).toEqual(null);
});

it('should remove a moderator in the community', async () => {
  // ensure user is removed as moderator from all channels
  const query = /* GraphQL */ `
    mutation removeCommunityModerator($input: RemoveCommunityModeratorInput!) {
      removeCommunityModerator (input: $input) {
        id
      }
    }
  `;

  const context = { user: owner };

  expect.assertions(5);
  const result = await request(query, { context, variables });
  expect(result).toMatchSnapshot();

  // ensure user is removed as moderator from community
  const getUsersCommunities = () =>
    db
      .table('usersCommunities')
      .getAll([input.userId, input.communityId], {
        index: 'userIdAndCommunityId',
      })
      .run();

  const communityConnections = await getUsersCommunities();
  expect(communityConnections).toHaveLength(1);
  expect(communityConnections[0].isModerator).toEqual(false);

  // ensure user is removed as moderator from all channels
  const getUsersChannels = async () => {
    const channelIds = await db
      .table('channels')
      .getAll(input.communityId, { index: 'communityId' })
      .map(row => row('id'))
      .run();

    return await db
      .table('usersChannels')
      .getAll(...channelIds, { index: 'channelId' })
      .filter({ userId: input.userId })
      .run();
  };

  const channelConnections = await getUsersChannels();
  expect(channelConnections).toHaveLength(1);
  expect(channelConnections.every(channel => !channel.isModerator)).toEqual(
    true
  );
});
