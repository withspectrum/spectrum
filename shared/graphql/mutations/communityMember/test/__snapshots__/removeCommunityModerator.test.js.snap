// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should remove a moderator in the community 1`] = `
Object {
  "data": Object {
    "removeCommunityModerator": Object {
      "id": "12803353-cb86-4109-8c4a-9abba79398d8",
    },
  },
}
`;
